Lemon does pain

LemonInTheDark opened the pull request:
This pr exists so I can test if adding extra plane masters will cause untenable clientside lag.
I will explain what it's doing later, I promise.
Thanks, love you, take care :)

LemonInTheDark committed:
Implements the plane cube backend

Ok so like, right now, we have an issue with multiz
Effects applied to planes displayed on lower z levels break when piped
through multiz.

This is because we use a vis_contents, and set everything's vis flags to
PARENT_PLANE.

This essentially tells the turfs and mobs we rerender to lose their
plane, and just render into the plane of the turf they're being copied
into.

This is typically mostly unoticable, but it absolutely borks side_map,
since it uses planes to separate things into uninterlayerable groups,
needs to since it does more advanced sorting based off the thing's
position.

So, the solution.

Instead of having just one set of plane masters, we have one for each
max "layer" of z interlocking in the game.

Rather then piping the lower layer's plane masters into the client's
view, we instead pipe them onto the openspace plane of the layer above.

This lets us apply effects to just X level, and also fixes this effects
failing issue.

I've attempted to implement this in such a way that reduces the impact
of the typical user. I will need to add better getters, and do a pass of
all uses of plane to either use float_plane, or to update on z change.

These should? be doable tho. I have hope

LemonInTheDark committed:
Does a pass on all /atom uses of plane, misc fixes

Converted all non hud (because you never actually want your hud elements
to underlay) of plane to the new system.
Ensures plane is properly set between moves
Adds hooks to track a client's eye plane offset, uses them to fix a bug
with the lighting plane
Adds a macro for getting all the planes in a group, uses it to make all
uses of plane master getting apply to all plane masters in a "type"
Adds support for plane masters rendering towards multiple targets. Plan
on using this in future
Shortens the plane master range, it was causing plane values to dip
below -10000, and brick the renderer
Makes some overlays "follow" their source, should really add a more
generic way of doing this

LemonInTheDark committed:
Some general fixes and cleanup

Fixes misplaned'd area lighting
Fixes new planes potentially being set to 0 if a move occured before
lists were built
Ensures parallax will only ever render on the bottom of a "z stack"
Fixes weird parallax behavior, turns out negative layers are "float"
layers, and were breaking it

LemonInTheDark committed:
Ensures the secure eye program does NOT attempt to load relay planes

LemonInTheDark committed:
Fixes some issues with excited group overlays

LemonInTheDark committed:
Adds a plane debugging ui

It lets you visualize how plane masters connect, edit their relays,
read details about each plane, and includes some in game documentation
for how plane masters work.

As a consequence of this, I've added a documentation variable on plane
masters. It forces devs to insert an explanation of what their plane
master does in text, partly so it can be displayed in text, and partly
just so it can be read.

Did some general plane cleanup, renamed backdrop() to show_to() because
backdrop() was NOT the right name, and the comment above it was HYPER
wrong.

LemonInTheDark committed:
Removes some unused code and cleans up some existing stuff

LemonInTheDark committed:
Adds offset support to mutable_appearance and emissives

It's actually only really needed when we manually set planes, which is
rare.
Need to do a pass on vis_contents and some other stuff, also uh, handle
on mob overlays. not sure what the heck I'ma do about that

LemonInTheDark committed:
Does a pass for underlays usage

Also cleans up some existing code, and ensures said existing code will
update on z move

LemonInTheDark committed:
Does a pass for vis_contents usage

I got most of them, but not the turf ones because I'm honestly not sure
how they'll work. I think I need to do some testing

Used vis_flags to apply inherit plane where appropriate

LemonInTheDark committed:
more todo

LemonInTheDark committed:
Haha johnathan I am dying (Resolves PM issues on map views)

Ok so like, there's this idea in byond, about submaps
Basically we can use a prefix before screen_loc values, and some byond
dmf fuckery to get a second map window to draw to.

We'd then setup some plane masters with appropriate screen locs, and use
vis_contents or just appearances to get an object with the appropriate
screen_loc to look like the thing we wanna mimic.

This is how we do camera consoles and such.

Basically, this system needs to account for offset planes. But as I was
working on it, it occured to me: shouldn't it also account for stuff
that modifies planes, basically shouldn't huds "know" what planes are
effecting them.

In persuit of this, I refactored how plane masters are handled, adding
plane groups, that represent planes rendered on some particular screen.

It in theory supports preloading plane masters, and applying them to
huds later, tho I have yet to find use for that.

In addition, life continues to be suffering, so I refactored all uses of
this pattern to use one consistent object, in the same rough way.

This makes the code a lot easier to reason about and read, and makes
making modifications to the pattern much easier

Note to self: If I'm gonna do this whole plane master group thing, I
might need a way to get existing "modifiacations" to plane masters

That or just nut up and make most things use the main group, not sure

Oh and I'm gonna make the debug window capable of editing these things,
for sniks

LemonInTheDark committed:
Adds support for viewing plane master groups to the ui

Disables it for now, because non main plane groups don't actually well,
work right now. They don't use relays, which is kinda the whole point of
the ui

Enabling it's a toggle tho, so hopefully it'll be easy to use in future

LemonInTheDark committed:
Subgroups are no longer used by effect sources, like controllers or say, lighting alpha

Too much risk of "dropping" a change, it's better this way
Oh and I documented a bunch of stuff, hopefully it'll help understanding

LemonInTheDark committed:
Cleans up map view logic a bit, makes gateways "work"

Ok so like, we can't do the old gateway effect cause vis planes aren't
being inherited
Anyway, instead, lets display that stuff in the ui, and bring back the
old effect for the portal being ready, give it a nice scanline effect

LemonInTheDark committed:
moth is bullied

LemonInTheDark committed:
Ripple doesn't do anything, ya yeeet

LemonInTheDark committed:
Adds checks for non existant spokesmen being passed into mut apper creation

Ensures limbs have spokesmen (most of the time)
Will need to speak with kapu about this

LemonInTheDark committed:
Adds logic to human overlays to allow for updating their planes "in place"

Ok so like, doing a full update would be too expensive, so instead we're
gonna create a list of all the overlays that either have a unique plane,
or have a child with a uniuqe plane

Then we walk up that list to remove overlays, and down it to add
Cause of the overlay flattening biz and all

Note: What I want to do in the end is store a copy of this list over
each iteration, rather then recalcing it each z change.

I can't atm tho, because when I try to I end up in fucky hell
Basically, overlays are weird, and there's something funky goin on.
I've seen behavior where appearances were pulled from the void, and
applied to the mob

Scares me

LemonInTheDark committed:
FUCK   YOU

LemonInTheDark committed:
Makes an attempt at including hud members into the plane thing

This'll need more work, but it's a start

LemonInTheDark committed:
Cleanliness, and makes the code compile

I'm gonna make huds not respect plane changes in future, but I want this
to be consistent so people don't fuck up by copying the wrong thing

LemonInTheDark committed:
Adds support for disabling offsets on some planes, also proper multiz darkness

Some planes really just shouldn't get stacked, mostly the hud related
ones

I've added a way to disable that in such a way that things attempting to
offset will just roll back to the main initial plane value
Useful.

I've also actually made the darkness plane matter, by adding the
see_blackness flag to all mobs
This lets me actually move around darkness, and shift it down or up
based on your z level.

LemonInTheDark committed:
cleans up some stuff, adds some todos, makes gas actually visible, pog

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
wahhh wahhh stop abusing macro- NO

LemonInTheDark committed:
fuck you

LemonInTheDark commented:
Note to self, I accidentially the hard deletes, need to fix that

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
Converts sight flags to a setter so I can react to them changing

This lets me get around blackness sticking around when it shouldn't
Oh also I've removed a lot of fluff from the blackness plane master
Doesn't seem to have needed it, and the blend mulitply broke slowly
changing alpha, which I'm not a fan of

LemonInTheDark committed:
Cleans up some sight sets

Also mandates see_blackness, cause you always need to be able to see it
for the pm to work.
Also removes a redundant sight flag thing from nightvision eyes

LemonInTheDark committed:
fuck off

LemonInTheDark committed:
wah wahhh

LemonInTheDark committed:
fixes c&d tests, adds some better errors, etc

LemonInTheDark committed:
fuck you again

LemonInTheDark committed:
Cleans up some todos, adds some new ones

LemonInTheDark commented:
Mooshimi also experienced some weird layering with catgirl ears, I think I may have borked some ordering, will need to look into it

LemonInTheDark committed:
cleans up some manual plane sets, I have a feeling this will be a pattern

Mooshimi commented:
The pink of cat ears dissapears if you travel z level and then is fixed with a sprite update like toggling mask

timothymtorres commented:
The blur effects is broken from #68074.  I'm guessing you forgot to add the plane to the list that creates blur effects.
 
I fixed this in an early PR in May by adding a missing plane to the plane masters list.  See: #67191

LemonInTheDark committed:
ais can see lighting properly, adds a client.eye setter, adds some suffering to prospective code

LemonInTheDark committed:
Adds support for editing another mob's planes, for debug purposes

Makes it a lot easier to solve "hey what's going on" to be able to use
the nice ui to change stuff, rather then you know, doing it manually
with vv

LemonInTheDark committed:
Adds a setter and signal for see_invis

LemonInTheDark committed:
Adds a setter for see_in_dark, signals too, simiar reason to see_invisible

LemonInTheDark committed:
Adds support for limited mirroring of the view of one mob onto another

Ties it into the plane debugger ui, to make debugging player issues on
live easier

It's quite hacky, mostly because of how plane masters work, but there's
really not another way at the moment

Sorry

LemonInTheDark committed:
Fixes blur not applying to the floor, I am in fact a dumbass

LemonInTheDark committed:
Adds support for enabling/disabling mirroring in the ui

LemonInTheDark commented:
> The blur effects is broken from #68074. I'm guessing you forgot to add the plane to the list that creates blur effects.

Nah I refactored plane master controllers and fatfingered a src.plane in there, so it was handing back 70 instances of gameworld

LemonInTheDark committed:
Removes all filtering from carbon updating logic

Overlay order effects rendering order, so if anything is filtered we run
the risk of dropping entries.
I might redo this at some point to drop inputs with no non floating
planes, but I feel like those will be rare so not right now

LemonInTheDark commented:
Fixed the ai thing, and ears not rendering
ai was simple, just eyes not being tracked, the ears was the result of overlay render order yorking, it's regrettably fixed now, but I can't filter so I'm sad

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
fine fine

LemonInTheDark committed:
bastards, fixes nonexisting parent calls

LemonInTheDark committed:
Adds support to multi cams, since they weren't working as expected

LemonInTheDark commented:
![image](https://user-images.githubusercontent.com/58055496/177231053-0fb9f354-a2e7-4827-80b2-dfb1eea8e6e1.png)
![image](https://user-images.githubusercontent.com/58055496/177231060-62412531-585c-437e-b7f5-b1c8f6cad676.png)
This is just lighting not being multiz, and the lower level being totally dark, here's me alphaing the lower lighting plane down a bit, can still see it

LemonInTheDark committed:
Ensures parallax works between z levels via mirage

Instead of positioning parallax objects on the lowest z level, we relay
them to all z levels
This way we don't need to move them, and they'll show up as expected

Note to self: add a way to edit relays and their blendmodes in future,
it'll be useful

Oh and I had to slightly rework how init relay adding is done, because
it didn't really support adding a relay with more data then just the
target plane
Not sure how I feel about this, I think it's alright though

LemonInTheDark committed:
haha get owned

LemonInTheDark committed:
Genericizes updating overlays, uses it to make statues work

LemonInTheDark committed:
Cleans up some appearance updates that weren't working great, added back a check for null locations being passed in

LemonInTheDark committed:
Moves emissive updating on init to hook into update_overlays

Overlays are not "like" removes like. if an overlay isn't added via
update_overlays, it can't be removed and refreshed automatically.

Emissive overlays were being added both in update_overlays, and in
initialize, which mean the one in initialize would never be removed

So every object that had one emissive blocker actually had two,
including the type that uses render targets.

sooooo lets just not do that one, and instead update appearance.

As a consequence, overrides of update_appearance and its subprocs need
to support being called at this stage in init, which many did not.

Oh and I've converted a lot of stuff that used the path -> datum in the
same var pattern, because it causes runtimes often with this method, and
is also really stupid

LemonInTheDark committed:
loosens the restrtictions on the mutable appearance check for planes, I wanna be nicer about this, needless mess for api users

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
whoops

LemonInTheDark committed:
wahhh wahh no var/ in proc args wahhh

timothymtorres commented on code/__HELPERS/lighting.dm:
```diff
@@ -1,11 +1,11 @@
 /// Produces a mutable appearance glued to the [EMISSIVE_PLANE] dyed to be the [EMISSIVE_COLOR].
-/proc/emissive_appearance(icon, icon_state = "", atom/offset_spokesman, layer = FLOAT_LAYER, alpha = 255, appearance_flags = NONE)
-	var/mutable_appearance/appearance = mutable_appearance(icon, icon_state, layer, offset_spokesman, EMISSIVE_PLANE, alpha, appearance_flags | EMISSIVE_APPEARANCE_FLAGS)
+/proc/emissive_appearance(icon, icon_state = "", atom/offset_spokesman, layer = FLOAT_LAYER, alpha = 255, appearance_flags = NONE, offset_const)
```
Lemon does this fix #65800

LemonInTheDark replied:
yes
I'll explain how when the pr's done

LemonInTheDark committed:
fixes up all the errors caused by the update_appearance in atom initialize

LemonInTheDark committed:
lmao lol

LemonInTheDark committed:
ensures weather overlays are properly mutliz. It's possible that this will lead to improper double layering, but I really don't want to make 3 extra planes just for this effect

LemonInTheDark commented:
note to self: the sand bug/persisting appearances is back again, you may have fucked it

LemonInTheDark committed:
enables blocking render relay creation, uses this to ensure there's only ever one fov plane master

LemonInTheDark committed:
tmp

LemonInTheDark committed:
Lays a protective ward of appearance.overlays.Copy()

Without this, appearances will "stick" and be impossible to remove.
They will also occasionally "steal" or corrupt into the appearance of
something else on the map. I do not know why.

I'm in the process of building a testcase for this but I want to make
sure it works fully first

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
fixes emissive overlays with no source

LemonInTheDark committed:
nother CI thing

LemonInTheDark committed:
sorts plane masters by plane, removes redundant byondlightingplane define, it was removed 4 years ago when ai camera static plane was created

LemonInTheDark commented:
> Paperbins have side effects in `update_overlays()` which gets called in `initialize` due to this PR ...which causes all paperbins to spawn empty

I love this place

LemonInTheDark committed:
jesus christ who mutates behavior in fucking update_overlays. Should also stop treating one var as two values, come on man

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
resolves runtimes in ci, cleans a mistake

LemonInTheDark committed:
I am dumb

iprice commented:
My linked issues (#68922 and #68923) were closed and I was told to comment them here.  Which the issue report kinda did its self, but I guess I need to note they're not actually resolved as the status implies, thats a consequence of them being closed, as fas as I know they still need resolving and you might as well just refer to the two now closed issues if you need the additional details.

LemonInTheDark commented:
Thanks. issue jannies like to keep the issue page as JUST current issues on master, with testmerge bugs commented on the particular pr. that's not always possible, but I already had my eye on those, will get to them (alongside my admin pm fuckup) hopefully tomorrow

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
Ensures FLOAT_PLANE is always properly managed by plane list mappings

As a consequence, resolves a bug with statue displaying
I added some extra code to make debugging it easier

Also solves a runtime in supermatter code

LemonInTheDark committed:
Fixes the podlauncher map not properly displaying

LemonInTheDark committed:
stacktraces in mirage destroy so I can see what's fucking the unit tests

LemonInTheDark committed:
WARNING WARNING TEMP

LemonInTheDark committed:
resolves random runtimes in unit tests

LemonInTheDark committed:
ahhhhhhhh

LemonInTheDark committed:
Cleans up SET_PLANE slightly, so most uses require less fluff

I'll document this soon

LemonInTheDark committed:
it is done

LemonInTheDark committed:
extra todo

LemonInTheDark committed:
Resolves a class of issue with hud images updating on loc move

Basically, we can handle images on basic huds updating loc generically
Since we have access to their loc on hud add, we can register off that,
and mange like that.

Also removes a redundant todo about ais and static through multiz, it's
an issue on master too

LemonInTheDark committed:
Resolves issues with updating some images on move. Nearly done with todos now

LemonInTheDark committed:
Removes redundant check. I don't wanna touch the rest of this until I'm ready for a fullmerge cause it'll be useful in debugging, so next is docs and then I'm gonna pr this thing

timothymtorres commented:
Check and see if this fixes #45669

LemonInTheDark commented:
It will not fix https://github.com/tgstation/tgstation/issues/45669
https://github.com/tgstation/tgstation/issues/50444 is not intended, and also my fault. We (somehow I still don't know what the actual cause was) fixed it for humans, so it ought to be trivial to fix for slimes

LemonInTheDark committed:
Adds documentation for EVERYTHING I know about our current rendering systems.

This includes my plane master changes, relays, appearances, and all the
little bits I think are useful to know/mention
